File: src/DashSpec.Core/Parsing/FiltersChromeParser.cs
using DashSpec.Core.Model;

namespace DashSpec.Core.Parsing;

internal static class FiltersChromeParser
{
    public static FiltersChromeDefinition Parse(TokenReader reader)
    {
        var props = PropertyBlockParser.Parse(reader, PropertySchemas.FiltersChrome, "filters chrome");

        var layout = "card";
        var sticky = FiltersChromeDefinition.StickyNone;
        var apply = "manual";
        var debounceMs = 400;

        if (props.TryGetValue("layout", out var layoutRaw))
        {
            layout = layoutRaw.ToLowerInvariant() switch
            {
                "card" or "bar" => layoutRaw.ToLowerInvariant(),
                _ => throw new DashSpecParseException("filters chrome layout must be 'card' or 'bar'."),
            };
        }

        if (props.TryGetValue("sticky", out var stickyRaw))
        {
            sticky = FiltersChromeStickyParser.Parse(stickyRaw);
        }

        if (props.TryGetValue("apply", out var applyRaw))
        {
            apply = applyRaw.ToLowerInvariant() switch
            {
                "manual" or "auto" => applyRaw.ToLowerInvariant(),
                _ => throw new DashSpecParseException("filters chrome apply must be 'manual' or 'auto'."),
            };
        }

        if (props.TryGetValue("debounce_ms", out var debounceRaw) &&
            int.TryParse(debounceRaw, out var parsedDebounce) &&
            parsedDebounce >= 0)
        {
            debounceMs = parsedDebounce;
        }

        return new FiltersChromeDefinition(layout, sticky, apply, debounceMs);
    }
}

File: src/DashSpec.Modeling.Parse/Filter/FiltersChromeParser.fs
namespace DashSpec.Modeling.Parse.Filter

open System
open DashSpec.Modeling.Core
open DashSpec.Modeling.Parse
open DashSpec.Modeling.Parse.Lexing

module FiltersChromeStickyParser =

    let parse (raw: string option) =
        match raw with
        | None -> FiltersChromeDefinition.StickyNone
        | Some value when String.IsNullOrWhiteSpace value -> FiltersChromeDefinition.StickyNone
        | Some value ->
            match value.Trim().ToLowerInvariant() with
            | "true" | "yes" | "1" -> FiltersChromeDefinition.StickyLine
            | "false" | "no" | "0" -> FiltersChromeDefinition.StickyNone
            | FiltersChromeDefinition.StickyNone -> FiltersChromeDefinition.StickyNone
            | FiltersChromeDefinition.StickyLine -> FiltersChromeDefinition.StickyLine
            | FiltersChromeDefinition.StickyCard -> FiltersChromeDefinition.StickyCard
            | _ ->
                raise (DashSpecParseException("filters chrome sticky must be 'none', 'line', or 'card' (true/false also accepted)."))

module FiltersChromeParser =

    let parse (reader: TokenReader) =
        let props =
            PropertyBlockParser.parseWithEndKind
                reader
                PropertySchemas.filtersChrome
                "filters chrome"
                "chrome"
                false
                false

        let layout =
            match props.TryGetValue "layout" with
            | true, layoutRaw ->
                match layoutRaw.ToLowerInvariant() with
                | "card" | "bar" -> layoutRaw.ToLowerInvariant()
                | _ -> raise (DashSpecParseException("filters chrome layout must be 'card' or 'bar'."))
            | false, _ -> "card"

        let sticky =
            match props.TryGetValue "sticky" with
            | true, stickyRaw -> FiltersChromeStickyParser.parse (Some stickyRaw)
            | false, _ -> FiltersChromeDefinition.StickyNone

        let apply =
            match props.TryGetValue "apply" with
            | true, applyRaw ->
                match applyRaw.ToLowerInvariant() with
                | "manual" | "auto" -> applyRaw.ToLowerInvariant()
                | _ -> raise (DashSpecParseException("filters chrome apply must be 'manual' or 'auto'."))
            | false, _ -> "manual"

        let debounceMs =
            match props.TryGetValue "debounce_ms" with
            | true, debounceRaw ->
                match Int32.TryParse debounceRaw with
                | true, parsed when parsed >= 0 -> parsed
                | _ -> 400
            | false, _ -> 400

        { Layout = layout; Sticky = sticky; Apply = apply; DebounceMs = debounceMs }
